### PR TITLE
Add Schema model to represent an Endpoint's JSON structure versioned over time.

### PR DESCRIPTION
**Pull request by mcw0933**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 KimonoRSS is a tool for easily consuming JSON APIs from [KimonoLabs](https://www.kimonolabs.com) either in a ruby environment or as an RSS feed of entry history.
 
+## Developing
+
+You will need Postgres 9.3 or higher.  [Get Postgres.app](http://postgresapp.com/) and then use [the guide](http://postgresapp.com/documentation/cli-tools.html) to get set up.
+
```
Comment (jguluarte) [inline README.md]: Totally optional, but I believe you can also add a line for installing via homebrew (if the system is a mac).  Homebrew++

File: app/models/schema.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  private
+
+  # Ideally we would do this with a database default, but ActiveRecord doesn't like that:
+
+  # ArgumentError: bad value for range
+  # from ... gems/activerecord-4.2.0/lib/active_record/connection_adapters/postgresql/oid/range.rb:39:in `initialize'
+  def default_values
+    self.effective ||= DateTime.new(1, 1, 1)..DateTime::Infinity.new
+  end
```
Comment (jguluarte) [inline app/models/schema.rb]: Um...what if you just allow null values?

Comment (jguluarte) [inline app/models/schema.rb]: I'm not actually sure if that'd work.  I just think this looks odd to me.  Not saying it is wrong...just that is looks weird.

File: spec/models/endpoint_spec.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe Endpoint, type: :model do
+  context 'when created' do
+    subject { described_class.new }
+
+    it 'requires a kimono_id' do
```
Comment (jguluarte) [inline spec/models/endpoint_spec.rb]: I think that I saw this in #20 - does this PR depend on that one?

File: spec/models/schema_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+RSpec.describe Schema, type: :model do
+  context 'when created' do
+    subject { described_class.new }
+
+    it 'must have an endpoint and a structure' do
+      expect { subject.save! }.to raise_error(ActiveRecord::RecordInvalid)
+      subject.structure = '{ "field": "value" }'
+      expect { subject.save! }.to raise_error(ActiveRecord::StatementInvalid)
+    end
```
Comment (jguluarte) [inline spec/models/schema_spec.rb]: In rspec, it is generally frowned upon to have multiple `expect` statements within one test unless it is *absolutely critical* that both pass for the test to succeed.  In this case, I think that you might want to change how this test file is structured.

File: spec/models/schema_spec.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    it 'must have a non-empty structure' do
+      subject.structure = '{ }'
+      expect { subject.save! }.to raise_error(ActiveRecord::RecordInvalid)
+    end
```
Comment (jguluarte) [inline spec/models/schema_spec.rb]: Since you have another test here that is also testing *both* errors, and structures...this is what I propose:

Comment (jguluarte) [inline spec/models/schema_spec.rb]: Since you have your overall `describe` for the test, and then you wrap everything inside one more context...there is no real reason for that except test output.  So why not change it to be like this:
```ruby
describe class do
  subject { described_class.new }
  describe "Schema errors correctly when" do
    it "does not have an endpoint" do
       expect { subject.save! }.to raise_error(ActiveRecord::RecordInvalid)
    end
    context "#structure" do
      it "catches invalid statements" do
         subject.structure = '{ "field": "value" }'
         expect { subject.save! }.to raise_error(ActiveRecord::StatementInvalid)
      end
      it "cannot have an empty structure" do
        subject.structure = '{ }'
        expect { subject.save! }.to raise_error(ActiveRecord::RecordInvalid)
      end
    end
  end
end
```
Now, that said...I don't like all of the test names I have suggested.  But, this gives you a little bit more future-flexibility of where to place things.  And you also need to still decide where to place the DateTime tests.

Comment (jguluarte) [inline spec/models/schema_spec.rb]: Also keep in mind, that it might not really be useful to test ActiveRecord.  Unless you have good reason to do so.

File: README.md (hunk 1 above)
Comment (mcw0933) [inline README.md]: You know, I originally installed Postgres via homebrew, and I would agree, but the [Heroku setup](https://devcenter.heroku.com/articles/getting-started-with-rails4) guide [recommends the app](https://devcenter.heroku.com/articles/heroku-postgresql#local-setup).  

It turned out that doing it on homebrew I had the 9.4 client tools talking to a 9.1.3 server instance.  Which totally nuked my plan to do range types with the Schema model.  Took me a good hour to figure that out because `which psql` and `psql -V` was (obvs) telling me 9.4.

I (sadly) did `brew uninstall postgres` and followed this and it worked like a charm.  

TL;DR - n00b issues, following docs :/

File: app/models/schema.rb (hunk 2 above)
Comment (mcw0933) [inline app/models/schema.rb]: The idea here is to leverage [Ruby range types using ActiveRecord and Postgres](http://edgeguides.rubyonrails.org/active_record_postgresql.html#range-types) so I can say things like `schema.effective.include?(Date.today)`, and leverage the other range operators.  I've dealt with periodic data like this in other projects and date ranges are fantastic for them.

That said, I'm getting the sense that [this isn't as done](https://github.com/rails/rails/issues/17288) as [I initially thought](https://github.com/rails/rails/issues/14010).  Going to play with it some more and see.

File: spec/models/endpoint_spec.rb (hunk 3 above)
Comment (mcw0933) [inline spec/models/endpoint_spec.rb]: Yes, I didn't want to wait for review on #20 so I went ahead with this one.

Comment (jguluarte) [inline spec/models/endpoint_spec.rb]: I can see value in keeping it separate, and this PR just depends on that one.  But ultimately it is your call.

File: spec/models/schema_spec.rb (hunk 5 above)
Comment (mcw0933) [inline spec/models/schema_spec.rb]: Yeah, I'm more trying to make sure that the invariants on my models hold.  I'm less worried about the relational integrity tests (e.g. has an endpoint) and more about the expectations on the JSON data and the effective `tsrange`.
